src/Storages/MergeTree/MergeTreeIndexTextPostingListCodec.cpp

```diff
@@ -0,0 +1,358 @@
+            compressed_data.reserve(compressed_data.size() + 2 * min_need);
+        }
+        /// Block Layout: [1byte(bits)][payload]
+        size_t offset = compressed_data.size();
+        compressed_data.resize(compressed_data.size() + cap + 1);
+        auto * p = reinterpret_cast<unsigned char *>(compressed_data.data() + offset);
+
+        encodeU8(bits, p);
+        auto used = BlockCodec::encode(segment.data(), segment.size(), bits, p);
+        chassert(used == cap);
+
+        segments.back().compressed_data_size = compressed_data.size() - segments.back().compressed_data_offset;
+        current.clear();
+    }
+
+    /// Decode one compressed block into `current` and reconstruct absolute docIDs.
+    ///
+    /// - Reads bits-width byte
+    /// - Codec::decode fills `current` with delta values
+    /// - inclusive_scan converts deltas -> docIDs using `prev_value` as initial prefix
+    /// - Updates prev_value to the last decoded docID
+    static void decodeOneBlock(unsigned char * & in, size_t count, uint32_t & prev_value, std::vector<uint32_t> & current)
+    {
+        uint8_t bits = decodeU8(in);
+        current.resize(count);
+
+        /// Decode postings to buffer named temp.
+        BlockCodec::decode(in, count, bits, current.data());
+
+        /// Restore the original array from the decompressed delta values.
+        std::inclusive_scan(current.begin(), current.end(), current.begin(), std::plus<uint32_t>{}, prev_value);
+        prev_value = current.empty() ? prev_value : current.back();
+    }
+
+    ALWAYS_INLINE static void encodeU8(uint8_t x, unsigned char * & out)
+    {
+        *out++ = static_cast<char>(x);
+    }
+
+    ALWAYS_INLINE static uint8_t decodeU8(unsigned char * & in)
+    {
+        return *in++;
+    }
+
+    std::string compressed_data;
+    uint32_t prev_value = {};
+
+    /// Number of values added in the current segment.
+    size_t total_in_current_segment = 0;
+    std::vector<uint32_t> current;
+    size_t posting_list_block_size = 1024 * 1024;
+    std::vector<SegmentDesc> segments;
+
+    /// Total number of postings added across all segments.
+    size_t cardinality = 0;
+};
+
+void PostingListCodec::decode(ReadBuffer & in, PostingList & postings)
+{
+    PostingListCodecImpl impl;
+    impl.deserialize(in, postings);
+}
+
+void PostingListCodec::encode(const PostingListBuilder & postings, size_t posting_list_block_size, TokenPostingsInfo & info, WriteBuffer & out)
+{
+    PostingListCodecImpl impl(posting_list_block_size);
+    if (postings.isSmall())
+    {
+        const auto & small = postings.getSmall();
+        size_t small_size = postings.size();
+        for (size_t i = 0; i < small_size; ++i)
+            impl.add(small[i]);
+    }
+    else
+    {
+        std::vector<uint32_t> posting_values;
+        posting_values.resize(postings.size());
+        const auto & large = postings.getLarge();
+        large.toUint32Array(posting_values.data());
```

Comment:
I am asking myself if we can save a copy from Roaring (`postings`) to the integer array here. I guess we can iterate the bitmap directly to calculate gaps.